exercise2/ado77/common/treeHelpers.mjs

```diff
@@ -0,0 +1,159 @@
+class node {
+
+    constructor(data) {
+        this._value = data;
+    }
+
+    get value() {
+        return this._value;
+    }
+
+    set value(value) {
+        this._value = value;
+    }
+
+    get leftSide() {
+        return this._lhs;
+    }
+
+    set leftSide(value) {
+        this._lhs = value;
+    }
+
+    get rightSide() {
+        return this._rhs;
+    }
+
+    set rightSide(value) {
+        this._rhs = value;
+    }
+};
+
+const insertNode = (currentNode, value) => {
+    if (currentNode == undefined) {
+        return new node(value);
+    }
+    if (value === currentNode.value) {
+        return;
+    }
+
+    if (value < currentNode.value) {
+        if (currentNode.leftSide == undefined) {
+            return currentNode.leftSide = new node(value);
+        }
+        insertNode(currentNode.leftSide, value);
+    } else {
+        if (currentNode.rightSide == undefined) {
+            return currentNode.rightSide = new node(value);
+        }
+
+        insertNode(currentNode.rightSide, value);
+    }
+}
+
+export const buildTree = (data) => {
+
+    let root = new node(data[0]);
+
+    for (let index = 1; index < data.length; index++) {
+        insertNode(root, data[index]);
+    }
+
+    return root;
+}
+
+export const findTreeWidthAndDepth = (currentNode, currentWidth = 0, currentDepth = 0, offset = {
+    min: 0,
+    max: 0,
+    depth: 0
+}) => {
+
+    // console.log(`node Value: ${currentNode.value}, current: ${current}, Offset: ${offset.min},${offset.max}`);
+
+    if (currentWidth < offset.min)
+        offset.min = currentWidth;
+
+    if (currentWidth > offset.max)
+        offset.max = currentWidth;
+
+    if (currentDepth > offset.depth)
+        offset.depth = currentDepth;
+
+    if (currentNode.leftSide != undefined) {
+        findTreeWidthAndDepth(currentNode.leftSide, currentWidth - 1, currentDepth + 1, offset);
+    }
+
+    if (currentNode.rightSide != undefined) {
+        findTreeWidthAndDepth(currentNode.rightSide, currentWidth + 1, currentDepth + 1, offset);
+    }
+
+    return {
+        width: (offset.max + -offset.min + 2),
+        depth: offset.depth + 1
+    };
+}
+
+const _padNode = (arrToFill, start, length, char) => {
```

Comment:
What's the convention to start a function with an underscore?

Reply:
It is the convention for "private" in JS... which has no real private anything...